torch/_inductor/scheduler.py

```diff
@@ -109,6 +109,43 @@ def fuse(node1: "BaseSchedulerNode", node2: "BaseSchedulerNode"):
     if node1.is_foreach() or node2.is_foreach():
         return ForeachKernelSchedulerNode.fuse(node1, node2)
     else:
+        _, (vars1, reduce1) = node1.group
+        _, (vars2, reduce2) = node2.group
+        if vars1 != vars2:
+            # Here we start with strict conditions which may be relaxed later
+            c1 = reduce1 == () and reduce2 == ()
+            c2 = math.prod(vars1) == math.prod(vars2)
+            c3 = len(vars1) == 1 or len(vars2) == 1
+            if c1 and c2 and c3:
+
+                def get_indexing_ranges_exprs(node):
+                    if isinstance(node, SchedulerNode):
+                        comp_buffer = node.node
+                        _, body, _ = comp_buffer.get_default_sizes_body()
+                        return body.var_ranges, [*body.indexing_exprs.values()]
+                    elif isinstance(node, FusedSchedulerNode):
+                        assert len(node.snodes) > 0
+                        # use the last scheduler node from the list as it has the most
+                        # relevant indexing expressions
+                        return get_indexing_ranges_exprs(node.snodes[-1])
+                    else:
+                        return None
+
+                node_to_recomp = node1 if len(vars1) < len(vars2) else node2
+                ref_node = node2 if len(vars1) < len(vars2) else node1
+                extra_indexing_constraints = get_indexing_ranges_exprs(ref_node)
+                if extra_indexing_constraints is not None and isinstance(node_to_recomp, SchedulerNode):
+                    node_to_recomp.recompute_size_and_body(extra_indexing_constraints=extra_indexing_constraints)
```

Comment:
One concern is that retracing the body is fairly expensive, so we'll need to benchmark this to see how it impacts compile time overhead. I don't expect it to be too bad though as scheduler time doesn't usually dominate compilation.